### PR TITLE
feat(mocktail)!: Type Safe API

### LINKED ISSUE #13: NoMoreInteractions like in Mockito
@felangel what do you think about NoMoreInteractions to be sure that no other methods were called?

### LINKED ISSUE #16: Future proofing
**Is your feature request related to a problem? Please describe.**
When we create a test double (stub, mock, other...) for a class which defines a method that returns a future, it is tempting to use Mocktail to return a canned future. This is a good approach to improve speed and reliability. Yet it must be arranged in a careful way to avoid complications. 

Mocktail should prohibit developers to return a canned future with `.thenReturn()`. Instead, developers should be advised to use `.thenAnswer()`. 

Indeed, as Futures don't have a "start button", they run soon after they are _instantiated_. In the context of a canned future for tests, they are unfortunately likely to complete before our production code asks for the future.

Since `await` doesn't yield/block when applied to a _completed_ future (it just unpacks the value), we write asynchronous code but run it synchronously from our tests. In short, we believe the production code is tested when we are actually testing a different program.

In my case, this led to many race conditions and an overcomplicated solution to address these race conditions.

**Describe the solution you'd like**
**My suggestion is to throw an error from within `.thenReturn()` when the argument is of type Future.** The error, may provide a brief explanation as to why it's a mistake, and a suggestion to **use `.thenAnswer()` instead.**

This is what I changed in a [fork](https://github.com/svarlet/mocktail) of your repository and have been running it successfully for a few days now. I did find that 42 out of ~150 tests were making this mistake. 

Provided that in rare cases it is perhaps desirable to return a completed Future, Mocktail could provide a distinct method to allow this, even though it should be discouraged. Perhaps the name of that method will leave no doubts?

**Describe alternatives you've considered**
None besides being careful and consistent.

**Additional context**
The docs of [SynchronousFuture](https://api.flutter.dev/flutter/foundation/SynchronousFuture-class.html) warn us but are not very well connected to your library:

> In general use of this class should be avoided as it is very difficult to debug such bimodal behavior.

I also assume that alternatives to Mocktail expose their users to a similar issue.

### LINKED ISSUE #17: Overhead on refactor of class
**Is your feature request related to a problem? Please describe.**
When I rename the method of a class that is mocked I have to manually update the tests. 

Lets assume this simple class:

```
abstract class A {
   B add(B object);
}
```

In a test the class would be mocked:
```
class AMock extends Mock implements A {}

void main () {
   late AMock api;

   setUp((){
      api = AMock();
   });

   test('...', () {
      final b = B();
      when(api).calls(#add).thenReturn(b.copyWith(id: 'unique_id_from_api'));
      ... test continues ...
   });
}
```

First, there is no IntelliSense for `#add` which could introduce typos. Second, if `add` would be renamed I need to manually fix the `#add` in the tests. 

**Describe the solution you'd like**
No real solution in my head right now. Just wanted to start a conversation on how to go forward with this as there is a need for an easy testing framework with null safety and besides this issue, I like the usability of mocktail!

Any thoughts on the topic?

### PR DESCRIPTION
**Pull request by felangel**
## Status

**IN DEVELOPMENT**

## Breaking Changes

YES

## Description

- feat(mocktail)!: Type Safe API (closes #17, closes #16, closes #13)

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [X] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [X] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/mocktail/lib/src/mocktail.dart [hunk 1]
```diff
@@ -1,664 +1,729 @@
...
+final _timer = _TimeStampProvider();
+final _capturedArgs = <dynamic>[];
+final _storedArgs = <ArgMatcher>[];
+final _storedNamedArgs = <String, ArgMatcher>{};
+final _verifyCalls = <_VerifyCall>[];
+
+/// Opt-into [Mock] throwing [NoSuchMethodError] for unimplemented methods.
+///
+/// The default behavior when not using this is to always return `null`.
+void throwOnMissingStub(
```
Comment (samandmoore) [inline packages/mocktail/lib/src/mocktail.dart:33]: I've always found it weird that mockito in dart returns null for unstubbed method calls on Mocks.

I forget if that's the Java behavior, maybe that's where that comes from, I suspect it is because I have vague memories of debugging NREs in Java in my tests because of missing stubs.

Other languages I've used don't take that approach: they require you to stub all methods you intend to call otherwise they throw. 

I'm wondering is if you have any interest in adding an option to make all Mocks throw on unstubbed calls? And I'm also wondering if that might remove the need for the fallback values for user types?

**Review by samandmoore — COMMENTED** [review]
This is exciting!

I've only just started reading the code here, but I'm super excited about this. Awesome work, Felix!

File: packages/mocktail/lib/src/mocktail.dart (hunk 1 above)
Comment (samandmoore) [reply packages/mocktail/lib/src/mocktail.dart:33]: Oh I think I figured out that the answer to my second question is no. You'd still need the fallback value for parameters, right?

Comment (felangel) [reply packages/mocktail/lib/src/mocktail.dart:33]: Yup and you can use throwOnMissingStub https://github.com/felangel/mocktail/blob/f5ac82ed45e0791eb6552414f674fab79bfd378a/packages/mocktail/lib/src/mocktail.dart#L33 👍

Comment (samandmoore) [reply packages/mocktail/lib/src/mocktail.dart:33]: Oh, yea I see I can use that method on an individual mock instance, but I'm just wondering if you'd be open to a global setting to default to that behavior instead of having to call it each time I set up a mock.

Comment (felangel) [reply packages/mocktail/lib/src/mocktail.dart:32]: Yeah I'm definitely in favor of that but it would be a breaking change for those migrating from mockito which is why I maintained the same behavior. wdyt?

Comment (samandmoore) [reply packages/mocktail/lib/src/mocktail.dart:32]: Ah yea. I wouldn't want to force everyone to have to change strategy just to migrate. But I could imagine making it a globally set table option that defaults to false. Like `Mocktail.throwOnUndefinedStub = true`. 

It could default to false OR if you want this to be the intended way to use the library then you could default it to true and encourage folks migrating from Mockito to start by turning it off.

Comment (felangel) [reply packages/mocktail/lib/src/mocktail.dart:31]: You can do `Mock.throwOnMissingStub()` to configure this globally as part of v0.1.0 🎉
